### ISSUE TITLE
(Client=55) Missing class file for AWS discovery: TcpIpJoinerOverAWS

### ISSUE THREAD

**Issue by thecyman** [orientechnologies/orientdb#3304]
Hi guys-

I'm using 2.0 M3. I'm testing on a cluster.

The class `TcpIpJoinerOverAWS` is not included in the default `hazelcast` JAR. I think you need to use the HazelCast JAR that includes everything: `hazelcast-all-3.3.3.jar` from the HazelCast ZIP download.

I solved this by:
1. Downloading HazelCast 3.3.3, and unzip
2. Replace the file `ORIENTDB_HOME/libs/hazelcast-3.3.jar` with this file from the zip: `hazelcast-all-3.3.3.jar`

NOTE: I haven't tested this yet! Is there a problem using HazelCast 3.3.3 vs HazelCast 3.3 for clustering? Let me know, thanks!

```
2014-12-31 01:02:43:659 INFO  [172.31.32.191]:2434 [orientdb-pumpup] [3.3] Hazelcast 3.3 (20140907) starting at Address[172.31.32.191]:2434 [system]
2014-12-31 01:02:43:688 INFO  [172.31.32.191]:2434 [orientdb-pumpup] [3.3] Copyright (C) 2008-2014 Hazelcast.com [system]
2014-12-31 01:02:43:689 INFO  [172.31.32.191]:2434 [orientdb-pumpup] [3.3] Creating AWSJoiner [Node][172.31.32.191]:2434 [orientdb-pumpup] [3.3] Error while creating AWSJoiner!
java.lang.ClassNotFoundException: com.hazelcast.cluster.TcpIpJoinerOverAWS
        at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:191)
        at com.hazelcast.instance.Node.createJoiner(Node.java:577)
        at com.hazelcast.instance.DefaultNodeContext.createJoiner(DefaultNodeContext.java:35)
        at com.hazelcast.instance.Node.<init>(Node.java:172)
        at com.hazelcast.instance.HazelcastInstanceImpl.<init>(HazelcastInstanceImpl.java:120)
        at com.hazelcast.instance.HazelcastInstanceFactory.constructHazelcastInstance(HazelcastInstanceFactory.java:153)
        at com.hazelcast.instance.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:136)
        at com.hazelcast.instance.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:112)
        at com.hazelcast.core.Hazelcast.newHazelcastInstance(Hazelcast.java:58)
        at com.orientechnologies.orient.server.hazelcast.OHazelcastPlugin.configureHazelcast(OHazelcastPlugin.java:1032)
        at com.orientechnologies.orient.server.hazelcast.OHazelcastPlugin.startup(OHazelcastPlugin.java:133)
        at com.orientechnologies.orient.server.OServer.registerPlugins(OServer.java:796)
        at com.orientechnologies.orient.server.OServer.activate(OServer.java:235)
        at com.orientechnologies.orient.server.OServerMain.main(OServerMain.java:36)

2014-12-31 01:02:43:856 INFO  [172.31.32.191]:2434 [orientdb-pumpup] [3.3] Address[172.31.32.191]:2434 is STARTING [LifecycleService]
2014-12-31 01:02:46:366 WARNING [172.31.32.191]:2434 [orientdb-pumpup] [3.3] No join method is enabled! Starting standalone. [Node]
2014-12-31 01:02:46:892 INFO  [172.31.32.191]:2434 [orientdb-pumpup] [3.3] Address[172.31.32.191]:2434 is STARTED [LifecycleService]
2014-12-31 01:02:47:933 INFO  [172.31.32.191]:2434 [orientdb-pumpup] [3.3] Initializing cluster partition table first arrangement... [InternalPartitionService]
2014-12-31 01:02:48:490 INFO  [node1] found no previous messages in queue orientdb.node.node1.response [OHazelcastDistributedMessageService]
```

**Comment by Briksins** [comment]
Confirming the same problem with OrientDB 2.0.6

Solution:
Download Hazelcast 3.4.2 from http://hazelcast.org/download/
Extract hazelcast-all-3.4.2.jar, renamed and moved to: $ORIENTDB_HOME/libs/hazelcast-3.3.jar

**Comment by lvca** [comment]
You can also put hazelcast aws jar in OrientDB's libs directory

**Comment by Briksins** [comment]
invalid? wontfix?
2 people reporting problem and giving solution how to fix it, all you need is to re-package your install zip with fresh Jar
How come its invalid? and why are you not planning to fix it?
Or you think its ok when customers after installation have to manually fix some dependancy libraries for your product? 
Scary attitude....

**Comment by lvca** [comment]
@Briksins I explained that using AWS you need hazelcast cloud jars in your classpath (look at: http://docs.hazelcast.org/docs/1.9/manual/html/ch11s02.html). I don't like the solution to put al dependencies by default even if you're not on AWS. It's much easier download the jar and put it under libs:

http://mvnrepository.com/artifact/com.hazelcast/hazelcast-cloud/3.3

The upcoming official distributed AMI we'll publish on AWS will already contain this dependency. In this case makes sense.

**Comment by lvca** [comment]
It was also documented under: http://www.orientechnologies.com/docs/last/Distributed-Configuration.html#cloud-support

**Comment by Briksins** [comment]
"I don't like the solution to put al dependencies by default even if you're not on AWS"
It is absolutely illogical!
1) You already include Hezelcast jar into your product, why not include "Hazelcast-all" jar instead of the shorter version which u use now?
2)  Quote from: http://docs.hazelcast.org/docs/3.1/manual/html/ch14s02.html#EC2AutoDiscovery
"You need to add hazelcast-cloud.jar dependency into your project" 
In this case OrientDB is that project which use Hazelcast! why me as customer have to worry about it?
It is absolutely illogical to have included short version of Hazelcast but advertise full range of features and ask user to manually include some other libraries dependencies to lib or path without providing any download links in docs.
3) As prove of real life scenario you have already 2 people reporting this problem as bug -  it is mean:
a. Your documentation is not clear
b. You make customers to waste a lot of time by searching for 3d party libraries and include it manually somewhere

Once again you have include Hazelcast jar into the build, it is just short version which causing problems and wasting people time, please replace it with full jar and it will solve all problems and save a lot of time to your customers. 
Dif in size between your cut jar and full jar is 4.2mb vs 4.7mb ~ 0.5mb total.

**Comment by lvca** [comment]
You convinced me ;-) Upgraded to HZ 3.3.5 ALL version.

**Comment by Briksins** [comment]
Thank you Sir! :)

P.S. Heavily testing your DB in cluster right now, soon will loge more tickets related to: ODistributedException
Node1 fail to sync with Node2. All Logs will be provided

**Comment by lvca** [comment]
@Briksins Please open new issues with the problem you find. Please use 2.0.8-SNAPSHOT or 2.1-SNAPSHOT for your tests. Both will be released this week.

**Comment by mohamnag** [comment]
isn't this solved on 2.0.x branch? I'm on 2.0.15 and still have this problem, I think when you say distributed and dont provide the hazelcast cloud lib, are just making fun of "distributed" word.

**Comment by lvca** [comment]
2.1 should have everything in it.
